### PR TITLE
Fix endpoints in StaticFileAuth sample 

### LINKED ISSUE #43274: src/security/samples/StaticFilesAuth always not working
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Hi, When I tried to run the sample in : src/Security/Sampels/StaticFilesAuth, look  like it was broken.  would you help us take a look?

when I tried to run the sample,  after login, access the url: /https://localhost:5001/MapAuthenticatedFiles

it willreport Exception:

```Exception
System.InvalidOperationException: The request reached the end of the pipeline without executing the endpoint: '/'. Please register the EndpointMiddleware using 'IApplicationBuilder.UseEndpoints(...)' if using routing.
   at Microsoft.AspNetCore.Builder.ApplicationBuilder.<>c.<Build>b__18_0(HttpContext context)
   at Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.StaticFiles.DirectoryBrowserMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.StaticFiles.DefaultFilesMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authorization.Policy.AuthorizationMiddlewareResultHandler.HandleAsync(RequestDelegate next, HttpContext context, AuthorizationPolicy policy, PolicyAuthorizationResult authorizeResult)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Builder.Extensions.MapMiddleware.InvokeCore(HttpContext context, PathString matchedPath, PathString remainingPath)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
```

I tried to debug into.
I found it will jump here:

```csharp
      app.Map("/MapAuthenticatedFiles", branch =>
        {
            branch.Use((context, next) => { SetFileEndpoint(context, files, null); return next(context); });
            branch.UseAuthorization();
            SetupFileServer(branch, files);
        });
```

the werid thing is: it will run into branch.Use( (context, next) => {} ), and then the pipeline was end, will not go through 

```csharp
           branch.UseAuthorization();
            SetupFileServer(branch, files);
```

would you help to take a look?

thanks

### Expected Behavior

it should work success

### Steps To Reproduce

_No response_

### Exceptions (if any)

System.InvalidOperationException: The request reached the end of the pipeline without executing the endpoint: '/'. Please register the EndpointMiddleware using 'IApplicationBuilder.UseEndpoints(...)' if using routing.
   at Microsoft.AspNetCore.Builder.ApplicationBuilder.<>c.<Build>b__18_0(HttpContext context)
   at Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.StaticFiles.DirectoryBrowserMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.StaticFiles.DefaultFilesMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authorization.Policy.AuthorizationMiddlewareResultHandler.HandleAsync(RequestDelegate next, HttpContext context, AuthorizationPolicy policy, PolicyAuthorizationResult authorizeResult)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Builder.Extensions.MapMiddleware.InvokeCore(HttpContext context, PathString matchedPath, PathString remainingPath)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)

### .NET Version

6.0.400

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by Tratcher**
Fixes #43274

Two changes broke this sample:
- Changing how static file middleware interact with endpoints: https://github.com/aspnet/Announcements/issues/488
- The Auth Resource is now an HttpContext, not an Endpoint. https://github.com/dotnet/aspnetcore/pull/22672/files

Putting this fix in main (8) since there's no product or test impact.

@hylinux

### CODE REVIEW

**Comment by DamianEdwards** [comment]
Interesting. Do you think the way this sample implements auth for static files is still idiomatic? Should we consider pointing to something like https://github.com/DamianEdwards/AspNetCorePathAuthorization instead?

**Comment by Tratcher** [comment]
The FallbackPolicy approach isn't granular, and the FileResult approach is too granular, you have to map out your whole file structure.

The path authorization approach is a reasonable compromise, are you planning to integrate that for 8?

**Comment by DamianEdwards** [comment]
To be clear, you're saying the approaches in the docs are problematic (too broad, too granular)?

The path authorization approach via metadata-only endpoints is workable in .NET 7+. We could consider adding first-class support for such endpoints in .NET 8 (perhaps avoiding the per-request `MatcherPolicy` and instead modifying the route graph nodes themselves) which would allow efficient path-based rules for AuthZ and possibly output caching too. I'll log an issue to cover that.

**Comment by Tratcher** [comment]
> To be clear, you're saying the approaches in the docs are problematic (too broad, too granular)?

I would call them limited, or not dynamic. Authorizing everything is ok if that's what you want. Authorizing a single known file is ok too. But those samples don't allow any in-between. E.g. you can't restrict who as access to which files without hard coding every file.

I wouldn't want this sample to depend on yours until it was integrated in, it's too much to code people would need to copy.

**Comment by Tratcher** [comment]
@DamianEdwards @halter73 Can you approve this fix reflecting the current reality and we'll revisit this sample after #43642?

**Review by DamianEdwards — APPROVED** [review]

File: src/Security/samples/StaticFilesAuth/Startup.cs [hunk 1]
```diff
@@ -134,12 +123,14 @@ private static void SetFileEndpoint(HttpContext context, PhysicalFileProvider fi
...
-            metadata.Add(new AuthorizeAttribute(policy));
+            var metadata = new List<object>
+            {
+                new DirectoryInfo(Path.GetDirectoryName(fileSystemPath)),
+                new AuthorizeAttribute(policy)
+            };
 
             var endpoint = new Endpoint(
-                c => throw new InvalidOperationException("Static file middleware should return file request."),
+                requestDelegate: null,
```
Comment (halter73) [inline src/Security/samples/StaticFilesAuth/Startup.cs:133]: This led me to this:

https://github.com/dotnet/aspnetcore/blob/2c557452385bb607da6bc62ba6c14c0a43190bdd/src/Middleware/StaticFiles/src/StaticFileMiddleware.cs#L94-L95

Gross! It's actually pretty clever though. Is this anywhere in our real docs?

**Review by halter73 — APPROVED** [review]

**Pull request merged**